modules/libcom/RTEMS/posix/rtems_init.c

```diff
@@ -0,0 +1,1159 @@
+/*************************************************************************\
+* Copyright (c) 2002 The University of Saskatchewan
+* EPICS BASE Versions 3.13.7
+* and higher are distributed subject to a Software License Agreement found
+* in file LICENSE that is included with this distribution.
+\*************************************************************************/
+/*
+ * RTEMS startup task for EPICS
+ *      Author: W. Eric Norum
+ *              eric.norum@usask.ca
+ *              (306) 966-5394
+ */
+
+/*
+ * This is the first implementation with posix and rtems5.
+ *  Currently only dhcp is supported with libbsd. The reading of the
+ *  environment variables from e.g. u-boot environment is still missing.
+ * Extensive tests so far only with qemu.
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <ctype.h>
+#include <time.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <termios.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/socket.h>
+#include <netinet/in.h>
+#include <arpa/inet.h>
+#include <assert.h>
+#include <rtems.h>
+#include <rtems/malloc.h>
+#include <rtems/error.h>
+#include <rtems/stackchk.h>
+#include <rtems/imfs.h>
+#include <rtems/shell.h>
+#include <pthread.h>
+#include <signal.h>
+#include <sched.h>
+#include <rtems/libio.h>
+#include <rtems/rtc.h>
+#include <time.h>
+#include <sys/unistd.h>
+
+#if __RTEMS_MAJOR__ > 4
+#include <rtems/tod.h>
+#else
+#include <rtems/score/tod.h>
+#endif
+
+#ifdef RTEMS_LEGACY_STACK
+#include <rtems/rtems_bsdnet.h>
+#include <librtemsNfs.h>
+#else // libbsd stack
+#include <rtems/bsd/bsd.h>
+#include <rtems/bsd/modules.h>
+#include <rtems/dhcpcd.h>
+#include <machine/rtems-bsd-commands.h>
+#endif
+
+#include <rtems/telnetd.h>
+#if __RTEMS_MAJOR__ > 4
+#include <rtems/printer.h>
+#endif
+#include "epicsVersion.h"
+#include "epicsThread.h"
+#include "epicsTime.h"
+#include "epicsExit.h"
+#include "envDefs.h"
+#include "errlog.h"
+#include "logClient.h"
+#include "osiUnistd.h"
+#include "iocsh.h"
+#include "osdTime.h"
+#include "epicsMemFs.h"
+#include "epicsEvent.h"
+#include "errlog.h"
+
+#include "epicsRtemsInitHooks.h"
+
+#ifdef RTEMS_LEGACY_STACK
+#pragma message "RTEMS uses legacy stack"
+#else
+#pragma message "RTEMS uses libbsd stack"
+// setup own ntp-client to get time
+#include "epicsNtp.h"
+epicsEventId 	dhcpDone;
+#endif
+
+/* these settings are needed by the rtems startup
+ * may provide by dhcp/bootp
+ * or environments from the "BIOS" like u-boot, motboot etc.
+ * these are settings within FHI infrastructure (01/13/2021)
+ */
+char rtemsInit_NTP_server_ip[16] = "141.14.142.121";
+char bootp_server_name_init[128] = "1001.1001@141.14.128.9:/Volumes/Epics";
+char bootp_boot_file_name_init[128] = "/Volumes/Epics/myExample/bin/RTEMS-beatnik/myExample.boot";
+char bootp_cmdline_init[128] = "/Volumes/Epics/myExample/iocBoot/iocmyExample/st.cmd";
```

Comment:
Need to remove this FHI site configuration.

Reply:
I removed the FHI details, and almost replaced with empty strings, but decided to keep a generic example instead, just so its clear what format the strings are expected to be in.

  99char rtemsInit_NTP_server_ip[16] = "127.0.0.1";
 100char bootp_server_name_init[128] = "1001.1001@127.0.0.1:/epics";
 101char bootp_boot_file_name_init[128] = "/epics/myExample/bin/RTEMS-beatnik/myExample.boot";
 102char bootp_cmdline_init[128] = "/epics/myExample/iocBoot/iocmyExample/st.cmd";

Reply:
Ack on the example, but I'd suggest not using the localhost IP which could be confusing. Maybe pick a non-routable address such as 192.168.5.1 or 10.0.5.1 say.

Reply:
I think this highlights the complications of how we are configuring RTEMS.  eg. the collection of weak symbols in rtems_init.c seems organic (and brittle).  The switch from RTEMS 4 -> 5 is an opportune time to straighten this out.

As the comment above this block says, this is getting overwritten from DHCP.  It should, but isn't, earlier overwritten from local non-volatile storage (cf. the ugly untested mess in setBootConfigFromNVRAM.c).  (order of decreasing preference: DHCP, NVRAM, application defaults, static defaults).  "application defaults" being a wishlist item for me.

Reply:
I tried to get the whole thing to work with DHCP so we can debug reasonably in qemu. I left the stuff with "ConfigFromNVRAM" aside and wanted to look at it again when I have more time. With U-Boot everything looks different too.